---
title: "Exploring NGS data - preliminary!"
output:
  html_notebook: default
  html_document:
    df_print: paged
  pdf_document: default
editor_options:
  chunk_output_type: inline
---

```{r setup_chunk}
library(data.table)
library(ggplot2)
library(gridExtra)

source("../custom_functions.R")
```

Data comes from a pipeline as described in the supplemental material from de Barba et al. (2016). It is from two HISEQ Illumina runs so it comes in two text files. The data is stored in `/DAB/data/` under `dab_hiseq1_genotypes.txt` and `dab_hiseq2_genotypes.txt`. A total of 2x 12 libraries have been analyzed.

```{r}
xy1 <- fread("./data/dab_hiseq1_genotypes.txt", stringsAsFactors = FALSE,
             colClasses = list(character = c(1, 4, 5, 7, 9, 11, 12),
                               numeric = c(2, 3, 6)))
xy2 <- fread("./data/dab_hiseq2_genotypes.txt", stringsAsFactors = FALSE,
             colClasses = list(character = c(1, 4, 5, 7, 9, 11, 12),
                               numeric = c(2, 3, 6)))

ngs1 <- list.files("./1_ngsfilters_hiseq1/", pattern = ".ngsfilter", full.names = TRUE)
ngs2 <- list.files("./1_ngsfilters_hiseq2/", pattern = ".ngsfilter", full.names = TRUE)
```


Number of unique samples in HISEQ1 is `r length(unique(xy1$Sample_Name))` and `r length(unique(xy2$Sample_Name))` in HISEQ2.

## Success of PCR
Has PCR succeeded. The data represents the number of sequences per plate. Plates are grouped together so that these plates represent four repeats of the same sample.

### HISEQ1
Plates 3, 4, 5, 6.
```{r}
sapply(sort(unique(xy1$Run_Name)), FUN = function(i, x){
  table(x[Run_Name == i & Plate %in% c(3, 4, 5, 6), ][order(Sample_Name, Marker, Plate, length)][, Plate])
}, x = xy1)
```

And plates 1, 2, 7 and 8.
```{r}
sapply(sort(unique(xy1$Run_Name)), FUN = function(i, x){
  table(x[Run_Name == i & Plate %in% c(1, 2, 7, 8), ][order(Sample_Name, Marker, Plate, length)][, Plate])
}, x = xy1)
```

### HISEQ2
Plates 3, 4, 5, 6.
```{r}
sapply(sort(unique(xy2$Run_Name)), FUN = function(i, x){
  table(x[Run_Name == i & Plate %in% c(3, 4, 5, 6), ][order(Sample_Name, Marker, Plate, length)][, Plate])
}, x = xy2)
```
Something has gone awry.

Plates 1, 2, 7, 8.
```{r}
sapply(sort(unique(xy2$Run_Name)), FUN = function(i, x){
  table(x[Run_Name == i & Plate %in% c(1, 2, 7, 8), ][order(Sample_Name, Marker, Plate, length)][, Plate])
}, x = xy2)
```

## Negative controls
Negative control samples are marked `AK` for tissue samples and `AC` for non-invasive samples. Some have designation `NeKo*`.

### HISEQ1
```{r}
xy1[grepl("(^AC\\..*$|^AK\\..*$|NeKo.*$)", Sample_Name), ][order(Sequence, Run_Name, Sample_Name, Plate, Marker)]
```

### HISEQ2
```{r}
xy2[grepl("(^AC\\..*$|^AK\\..*$|NeKo.*$)", Sample_Name), ][order(Sequence, Run_Name, Sample_Name, Plate, Marker)]
```


## Find blanks

List all samples which have a word `"blank"` in them.

### HISEQ1
```{r}
xy1[grepl("blank", Sample_Name), ]
```

### HISEQ2
```{r}
xy2[grepl("blank", Sample_Name), ]
```

## Number of reads per position
These figures should give you a broad idea of how many reads each position got. This may indicate weak samples. Also notice if there are reads in positions designated as blanks.

### HISEQ1
Results should be in `"./data/countplate"`
```{r eval = FALSE}
showSumsByLibrary(xy1, loc = "./data/", pos = "by_sample")
```

### HISEQ2
```{r eval = FALSE}
showSumsByLibrary(xy2, loc = "./data/", pos = "by_sample")
```

## Sequence frequency

For each locus, find sequences by frequency for two HISEQ runs.

### HISEQ1
```{r}
freq.of.seq.by.marker <- xy1[, .N, by = .(Marker, Sequence)][order(-Marker, N, decreasing = TRUE), .(Marker, N, Sequence)]
fwrite(freq.of.seq.by.marker, file = "./data/frequency_of_sequences_by_marker_hiseq1.txt", sep = "\t")
freq.of.seq.by.marker
```

### HISEQ2
```{r}
freq.of.seq.by.marker <- xy2[, .N, by = .(Marker, Sequence)][order(-Marker, N, decreasing = TRUE), .(Marker, N, Sequence)]
fwrite(freq.of.seq.by.marker, file = "./data/frequency_of_sequences_by_marker_hiseq2.txt", sep = "\t")
freq.of.seq.by.marker
```

## Distribution of flags

### HISEQ1
```{r}
table(xy1$flag)
```

### HISEQ2
```{r}
table(xy2$flag)
```

## Alleles and stutters
See the distribution of number of reads of alleles and stutters, drawn by marker.
### Alleles
```{r fig.width = 8, fig.height = 12}
pxy1 <- ggplot(xy1[called == TRUE], aes(x = Read_Count)) +
  geom_histogram() +
  xlab("Read count HISEQ1") +
  scale_x_continuous(limits = c(0, quantile(xy1$Read_Count[xy1$called == TRUE], 0.99))) +
  facet_wrap(~ Marker, ncol = 2)
pxy2 <- ggplot(xy2[called == TRUE], aes(x = Read_Count)) +
  geom_histogram() +
  xlab("Read count HISEQ2") +
  scale_x_continuous(limits = c(0, quantile(xy2$Read_Count[xy2$called == TRUE], 0.99))) +
  facet_wrap(~ Marker, ncol = 2)
grid.arrange(pxy1, pxy2, ncol = 2)
```
### Stutters
```{r fig.width = 8, fig.height = 12}
pxy1 <- ggplot(xy1[stutter == TRUE], aes(x = Read_Count)) +
  geom_histogram() +
  xlab("Read count HISEQ1") +
  scale_x_continuous(limits = c(0, quantile(xy1$Read_Count[xy1$stutter == TRUE], 0.99)),
                     breaks = c(0, 500, 1000)) +
  scale_y_continuous(breaks = c(0, 6000)) +
  facet_wrap(~ Marker, ncol = 2)
pxy2 <- ggplot(xy2[stutter == TRUE], aes(x = Read_Count)) +
  geom_histogram() +
  xlab("Read count HISEQ2") +
  scale_x_continuous(limits = c(0, quantile(xy2$Read_Count[xy2$stutter == TRUE], 0.99)),
                     breaks = c(0, 500, 1000)) +
  scale_y_continuous(breaks = c(0, 6000)) +
  facet_wrap(~ Marker, ncol = 2)
grid.arrange(pxy1, pxy2, ncol = 2)
```
```{r}
```

## Number of replications

To calculate quality index, number of replications is essential.
```{r}
fwrite(countSampleLocusRepeats(ngs = ngs1), file = "./data/dab_number_of_repeats_hiseq1.txt", sep = "\t")
fwrite(countSampleLocusRepeats(ngs = ngs2), file = "./data/dab_number_of_repeats_hiseq2.txt", sep = "\t")
```

## Create a list of samples
Map samples to their position in library. This will help us know which samples worked and which did not.

```{r}
outngs1 <- sapply(ngs1, FUN = sampleMetadata, simplify = FALSE)
outngs1 <- rbindlist(outngs1)
outngs2 <- sapply(ngs2, FUN = sampleMetadata, simplify = FALSE)
outngs2 <- rbindlist(outngs2)
out <- rbindlist(list(outngs1, outngs2))
fwrite(out, file = "./data/sample_positions_DAB01-DAB24.txt", sep = "\t")
```

# Find bad samples for hiseq1
```{r}
ngslib1 <- data.frame(
  path = ngs1, 
  lib = sprintf("DAB%s", gsub("^DAB(\\d+)\\.ngsfilter$", "\\1", basename(ngs1))))

dud <- findUnamplifiedSamples(x = xy1, ngs = ngslib1, sex = "./data/dab_called_genetic_sex_hiseq1.txt")

dud.nrow <- sapply(dud, nrow)
dud.nrow[order(names(dud.nrow))]

dud <- rbindlist(dud)
dud$length <- -1
dud$Sequence <- "-1"
fwrite(dud, file = "./data/bad_samples_hiseq1.txt", sep = "\t")
```

# Find bad samples for hiseq2
```{r}
ngslib2 <- data.frame(
  path = ngs2, 
  lib = sprintf("DAB%s", gsub("^DAB(\\d+)\\.ngsfilter$", "\\1", basename(ngs2))))

dud <- findUnamplifiedSamples(x = xy2, ngs = ngslib2, sex = "./data/dab_called_genetic_sex_hiseq2.txt")

dud.nrow <- sapply(dud, nrow)
dud.nrow[order(names(dud.nrow))]

dud <- rbindlist(dud)
dud$length <- -1
dud$Sequence <- "-1"
fwrite(dud, file = "./data/bad_samples_hiseq2.txt", sep = "\t")
```
